aten/src/ATen/native/layer_norm.cpp

```diff
@@ -0,0 +1,647 @@
+      1LL,
+      std::multiplies<std::int64_t>());
+  const std::int64_t N = std::accumulate(
+      X_dims.cbegin() + axis,
+      X_dims.cend(),
+      1LL,
+      std::multiplies<std::int64_t>());
+
+  const T* ddX_data = ddX.defined() ? ddX.template data<T>() : nullptr;
+  const T* ddgamma_data =
+      ddgamma.defined() ? ddgamma.template data<T>() : nullptr;
+  const T* ddbeta_data = ddbeta.defined() ? ddbeta.template data<T>() : nullptr;
+  const T* dY_data = dY.template data<T>();
+  const T* X_data = X.template data<T>();
+  const T* mean_data = mean.template data<T>();
+  const T* rstd_data = rstd.template data<T>();
+  const T* gamma_data = gamma.defined() ? gamma.template data<T>() : nullptr;
+  Tensor ddY;
+  Tensor dX;
+  Tensor dgamma;
+  T* ddY_data = nullptr;
+  T* dX_data = nullptr;
+  T* dgamma_data = nullptr;
+  if (grad_input_mask[0]) {
+    ddY = at::empty_like(dY);
+    ddY_data = ddY.template data<T>();
+  }
+  if (grad_input_mask[1]) {
+    dX = at::empty_like(X);
+    dX_data = dX.template data<T>();
+  }
+  if (grad_input_mask[2]) {
+    dgamma = at::zeros_like(gamma);
+    dgamma_data = dgamma.template data<T>();
+  }
+  const T scale = T(1) / static_cast<T>(N);
+  for (std::int64_t i = 0; i < M; ++i) {
+    const T* ddX_ptr = ddX_data == nullptr ? nullptr : ddX_data + i * N;
+    const T* dY_ptr = dY_data + i * N;
+    const T* X_ptr = X_data + i * N;
+    T s_ddx_dy;
+    T s_ddx_x;
+    T s_dy_x;
+    T s_ddx;
+    T s_dy;
+    ComputeInternalGradients<T>(
+        N,
+        ddX_ptr,
+        dY_ptr,
+        X_ptr,
+        gamma_data,
+        &s_ddx_dy,
+        &s_ddx_x,
+        &s_dy_x,
+        &s_ddx,
+        &s_dy);
+    if (ddY_data != nullptr) {
+      LayerNormOutputDoubleBackward<T>(
+          N,
+          mean_data[i],
+          rstd_data[i],
+          s_ddx_x,
+          s_ddx,
+          ddX_ptr,
+          ddgamma_data,
+          ddbeta_data,
+          gamma_data,
+          X_ptr,
+          ddY_data + i * N);
+    }
+    if (dX_data != nullptr) {
+      LayerNormInputDoubleBackward<T>(
+          N,
+          mean_data[i],
+          rstd_data[i],
+          s_ddx_dy,
+          s_ddx_x,
+          s_dy_x,
+          s_ddx,
+          s_dy,
+          ddX_ptr,
+          ddgamma_data,
+          dY_ptr,
+          X_ptr,
+          dX_data + i * N);
+    }
+    if (dgamma_data != nullptr) {
+      GammaDoubleBackward<T>(
+          N,
+          mean_data[i],
+          rstd_data[i],
+          s_ddx_x,
+          s_ddx,
+          ddX_ptr,
+          dY_ptr,
+          X_ptr,
+          dgamma_data);
+    }
+  }
+
+  return std::make_tuple(ddY, dX, dgamma);
+}
+
+} // namespace
+
+std::tuple<Tensor, Tensor, Tensor> LayerNormForwardCPU(
+    const Tensor& X,
+    const Tensor& gamma /* optional */,
+    const Tensor& beta /* optional */,
+    const IntArrayRef normalized_shape,
+    const double eps) {
+  const int normalized_ndim = normalized_shape.size();
+  AT_CHECK(
```

Comment:
wait, this is the same error checking code as above, but it just uses different variable names?  That is just unnecessarily confusing, no?